arwen/hosttest/contracts_deploy_test.go

```diff
@@ -0,0 +1,466 @@
+	testcommon "github.com/ElrondNetwork/arwen-wasm-vm/v1_4/testcommon"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/stretchr/testify/require"
+)
+
+var sc1Address = testcommon.MakeTestSCAddress("sc1")
+var sc2Address = testcommon.MakeTestSCAddress("sc2")
+
+type deployFromSourceTestConfig struct {
+	deployedContractAddress []byte
+	gasUsedByInit           uint64
+	gasProvided             uint64
+	gasProvidedForInit      uint64
+	asyncCallStepCost       uint64
+	aoTPreparePerByteCost   uint64
+	compilePerByteCost      uint64
+}
+
+type updateFromSourceTestConfig struct {
+	deployFromSourceTestConfig
+	contractToBeUpdatedAddress []byte
+	owner                      []byte
+	isFlagEnabled              bool
+	hasCallback                bool
+	callbackFails              bool
+}
+
+func getDeployFromSourceTestConfig() deployFromSourceTestConfig {
+	return deployFromSourceTestConfig{
+		deployedContractAddress: sc1Address,
+		gasUsedByInit:           uint64(200),
+		gasProvidedForInit:      uint64(300),
+		gasProvided:             uint64(1000),
+		aoTPreparePerByteCost:   uint64(1),
+		compilePerByteCost:      uint64(2),
+	}
+}
+
+func getUpdateFromSourceTestConfig() updateFromSourceTestConfig {
+	return updateFromSourceTestConfig{
+		deployFromSourceTestConfig: getDeployFromSourceTestConfig(),
+		contractToBeUpdatedAddress: sc2Address,
+		owner:                      test.ParentAddress,
+		isFlagEnabled:              true,
+		hasCallback:                true,
+		callbackFails:              false,
+	}
+}
+
+// GetGasUsedByChild
+func (config deployFromSourceTestConfig) GetGasUsedByChild() uint64 {
+	return config.gasUsedByInit
+}
+
+// GetGasUsedByChild
+func (config updateFromSourceTestConfig) GetGasUsedByChild() uint64 {
+	return config.gasUsedByInit
+}
+
+// CallbackFails
+func (config updateFromSourceTestConfig) CallbackFails() bool {
+	return config.callbackFails
+}
+
+func TestDeployFromSource_Success(t *testing.T) {
+	testConfig := getDeployFromSourceTestConfig()
+	deployedCode := testConfig.deployedContractAddress /* this is the actual mock code of the deployed contract */
+	deployedCodeLen := uint64(len(deployedCode))
+	runDeployFromSourceTest(t, &testConfig, func(world *worldmock.MockWorld, verify *test.VMOutputVerifier) {
+		newContractAddress := verify.VmOutput.ReturnData[0]
+		verify.
+			Ok().
+			Code(newContractAddress, deployedCode).
+			GasRemaining(testConfig.gasProvided -
+				testConfig.gasUsedByInit -
+				deployedCodeLen*testConfig.compilePerByteCost -
+				deployedCodeLen*testConfig.aoTPreparePerByteCost)
+	})
+}
+
+func TestDeployFromSource_NoGasForInit(t *testing.T) {
+	testConfig := getDeployFromSourceTestConfig()
+	testConfig.gasProvidedForInit = uint64(100)
+	runDeployFromSourceTest(t, &testConfig, func(world *worldmock.MockWorld, verify *test.VMOutputVerifier) {
+		verify.
+			ExecutionFailed().
+			HasRuntimeErrors(arwen.ErrInputAndOutputGasDoesNotMatch.Error())
```

Comment:
So this doesn't fail with `OutOfGas`, but with `ExecutionFailed` due to gas mismatch? This is safe, not a problem, but must be investigated, this should not be a gas mismatch. Maybe leave a TODO?

Reply:
I think integrating the condition I mentioned in `v1_4_upgradeContract()` will ensure the final error will be the proper `OutOfGas`. With the epoch flag, of course.